File: Scala/src/main/scala/TextbookTrack/BA1L.scala
package TextbookTrack

object BA1L {
  private val NucleotideOrder: Map[Char, Int] = Map('A' -> 0, 'C' -> 1, 'G' -> 2, 'T' -> 3)

  def patternToNumber(pattern: String): Int =
    pattern.foldLeft(0)((acc, nucleotide) => 4 * acc + NucleotideOrder(nucleotide))

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val dna: String = reader.next()
    val result: Long = patternToNumber(dna)
    println(result)
  }
}

File: Python/textbook_track/chapter01/ba1l.py
# Implement PatternToNumber
import sys

NUCLEOTIDE_ORDER = {'A': 0, 'C': 1, 'G': 2, 'T': 3}


def pattern_to_number(pattern):
    encoding = 0
    for nucleotide in pattern:
        encoding = 4 * encoding + NUCLEOTIDE_ORDER[nucleotide]
    return encoding


def main():
    data = sys.stdin.read().splitlines()
    dna = data[0]
    result = pattern_to_number(dna)
    print(result)


if __name__ == '__main__':
    main()
